File: diffwake/diffwake_jax/flow_field.py
import jax
import jax.numpy as jnp
from flax import struct
from typing import Optional, Any


@struct.dataclass
class FlowField:
    wind_speeds: jnp.ndarray        
    wind_directions: jnp.ndarray    
    wind_shear: float
    wind_veer: float
    air_density: float
    turbulence_intensities: jnp.ndarray 
    reference_wind_height: float

    u_initial_sorted: Optional[jnp.ndarray] = None
    v_initial_sorted: Optional[jnp.ndarray] = None
    w_initial_sorted: Optional[jnp.ndarray] = None
    dudz_initial_sorted: Optional[jnp.ndarray] = None

    u_sorted:  Optional[jnp.ndarray] = None
    v_sorted:  Optional[jnp.ndarray] = None
    w_sorted:  Optional[jnp.ndarray] = None

    u: Optional[jnp.ndarray] = None
    v: Optional[jnp.ndarray] = None
    w: Optional[jnp.ndarray] = None
    dudz: Optional[jnp.ndarray] = None

    turbulence_intensity_field_sorted: Optional[jnp.ndarray] = None
    turbulence_intensity_field_sorted_avg: Optional[jnp.ndarray] = None

    turbulence_intensity_field: Optional[jnp.ndarray] = None
    grid_resolution: Optional[int] = None
    n_turbines:     Optional[int] = None
    n_findex:       Optional[int] = None
    state: str = "UNINITIALIZED"

    def initialize_velocity_field(self, grid: Any) -> "FlowField":
        """
        grid must expose:
            .z_sorted        : (B,T,Ny,Nz)
            .unsorted_indices: (B,T,Ny,Nz)  or (B,T,1,1) – same shape you used
        Returns a **new** FlowField with all *_sorted fields filled.
        """
        z_sorted = grid.z_sorted
        B, T, Ny, Nz = z_sorted.shape
        safe_z = jnp.clip(z_sorted, min=1e-7)

        wind_profile_plane = (safe_z / self.reference_wind_height) ** self.wind_shear
        dudz_profile = (
            self.wind_shear *
            (1.0 / self.reference_wind_height) ** self.wind_shear *
            safe_z ** (self.wind_shear - 1)
        )

        wind_speeds_exp = self.wind_speeds[:, None, None, None].astype(self.wind_speeds.dtype)  # (B,1,1,1)

        u_init = wind_speeds_exp * wind_profile_plane
        u_init = u_init.astype(self.wind_speeds.dtype)
        dudz   = wind_speeds_exp * dudz_profile
        dudz   = dudz.astype(self.wind_speeds.dtype)
        zeros  = jnp.zeros_like(u_init, dtype=self.wind_speeds.dtype)

        idxer  = grid.unsorted_indices
        u_uns  = jnp.take_along_axis(u_init, idxer, axis=1)
        v_uns  = jnp.take_along_axis(zeros,   idxer, axis=1)
        w_uns  = jnp.take_along_axis(zeros,   idxer, axis=1)
        dudz_uns = jnp.take_along_axis(dudz,   idxer, axis=1)
        # Expand TI field
        turb_exp = self.turbulence_intensities[:, None, None, None]  # (B,1,1,1)
        turb_exp = jnp.broadcast_to(turb_exp, (B, T, 1, 1))

        new_flow = self.replace(
            u_initial_sorted=u_init,
            v_initial_sorted=zeros,
            w_initial_sorted=zeros,
            dudz_initial_sorted=dudz,
            u_sorted=u_init,
            v_sorted=zeros,
            w_sorted=zeros,
            dudz = dudz_uns,
            u=u_uns, v=v_uns, w=w_uns,
            turbulence_intensity_field_sorted=turb_exp,
            turbulence_intensity_field=turb_exp,
            grid_resolution=Ny,
            n_turbines=T,
            n_findex=B,
            state="INITIALIZED",
        )

        return new_flow

    def finalize(self, unsorted_indices: jnp.ndarray) -> "FlowField":
        idx = unsorted_indices
        u_uns = jnp.take_along_axis(self.u_sorted, idx, axis=1)
        v_uns = jnp.take_along_axis(self.v_sorted, idx, axis=1)
        w_uns = jnp.take_along_axis(self.w_sorted, idx, axis=1)

        ti_uns = jnp.take_along_axis(
            self.turbulence_intensity_field_sorted, idx, axis=1
        )
        ti_field = jnp.mean(ti_uns, axis=(2, 3))

        return self.replace(u=u_uns, v=v_uns, w=w_uns,
                            turbulence_intensity_field=ti_field,
                            state="FINALIZED")

    def to_device(self, device) -> "FlowField":
        """
        device: 'cpu', 'gpu', 'tpu', or jax.Device
        Returns a copy with every ndarray placed on *device*.
        """
        def _move(x):
            return jax.device_put(x, device) if isinstance(x, jnp.ndarray) else x
        return self.map(_move)             # flax.struct dataclass utility

    @property
    def coordinates(self) -> jnp.ndarray:
        return jnp.stack(
            [self.layout_x, self.layout_y,
             jnp.full_like(self.layout_x, self.reference_wind_height)],
            axis=-1
        )

File: diffwake/diffwake_julia/flow_field.jl
"""
    FlowField

Atmospheric / flow-field state container.  Mirrors the Flax `FlowField`
dataclass from `flow_field.py`.
"""
Base.@kwdef mutable struct FlowField
    wind_speeds::Vector{Float64}
    wind_directions::Vector{Float64}
    wind_shear::Float64
    wind_veer::Float64
    air_density::Float64
    turbulence_intensities::Vector{Float64}
    reference_wind_height::Float64

    u_initial_sorted::Union{Nothing, Array{Float64,4}}      = nothing
    v_initial_sorted::Union{Nothing, Array{Float64,4}}      = nothing
    w_initial_sorted::Union{Nothing, Array{Float64,4}}      = nothing
    dudz_initial_sorted::Union{Nothing, Array{Float64,4}}   = nothing

    u_sorted::Union{Nothing, Array{Float64,4}}  = nothing
    v_sorted::Union{Nothing, Array{Float64,4}}  = nothing
    w_sorted::Union{Nothing, Array{Float64,4}}  = nothing

    u::Union{Nothing, Array{Float64,4}}     = nothing
    v::Union{Nothing, Array{Float64,4}}     = nothing
    w::Union{Nothing, Array{Float64,4}}     = nothing
    dudz::Union{Nothing, Array{Float64,4}}  = nothing

    turbulence_intensity_field_sorted::Union{Nothing, Array{Float64,4}}     = nothing
    turbulence_intensity_field_sorted_avg::Union{Nothing, Array{Float64,4}} = nothing
    turbulence_intensity_field::Union{Nothing, Array{Float64,4}}            = nothing

    grid_resolution::Union{Nothing, Int} = nothing
    n_turbines::Union{Nothing, Int}      = nothing
    n_findex::Union{Nothing, Int}        = nothing
end

"""
    initialize_velocity_field!(ff::FlowField, grid::TurbineGrid)

Fills `ff` with the initial velocity field using a power-law shear profile.
"""
function initialize_velocity_field!(ff::FlowField, grid::TurbineGrid)
    z_sorted = grid.z_sorted
    B, T, Ny, Nz = size(z_sorted)
    safe_z = max.(z_sorted, 1e-7)

    wind_profile_plane = (safe_z ./ ff.reference_wind_height) .^ ff.wind_shear
    dudz_profile = (
        ff.wind_shear .*
        (1.0 / ff.reference_wind_height) ^ ff.wind_shear .*
        safe_z .^ (ff.wind_shear - 1.0)
    )

    # Expand wind speeds: (B,) → (B,1,1,1)
    ws = reshape(ff.wind_speeds, B, 1, 1, 1)

    u_init = ws .* wind_profile_plane
    dudz   = ws .* dudz_profile
    zeros4 = zeros(B, T, Ny, Nz)

    # Un-sort for the inertial-frame copies
    idxer = grid.unsorted_indices
    u_uns    = _take_along_axis2(u_init, idxer)
    v_uns    = _take_along_axis2(zeros4, idxer)
    w_uns    = _take_along_axis2(zeros4, idxer)
    dudz_uns = _take_along_axis2(dudz,   idxer)

    # TI field (B, T, 1, 1)
    turb_exp = zeros(B, T, 1, 1)
    for b in 1:B
        turb_exp[b, :, 1, 1] .= ff.turbulence_intensities[b]
    end

    ff.u_initial_sorted      = u_init
    ff.v_initial_sorted      = copy(zeros4)
    ff.w_initial_sorted      = copy(zeros4)
    ff.dudz_initial_sorted   = dudz

    ff.u_sorted = copy(u_init)
    ff.v_sorted = copy(zeros4)
    ff.w_sorted = copy(zeros4)

    ff.u    = u_uns
    ff.v    = v_uns
    ff.w    = w_uns
    ff.dudz = dudz_uns

    ff.turbulence_intensity_field_sorted = turb_exp
    ff.turbulence_intensity_field        = copy(turb_exp)

    ff.grid_resolution = Ny
    ff.n_turbines      = T
    ff.n_findex        = B

    return ff
end

"""
    finalize!(ff::FlowField, unsorted_indices)

Un-sort the solved fields back to the original turbine ordering.
"""
function finalize!(ff::FlowField, unsorted_indices::Array{Int,4})
    ff.u = _take_along_axis2(ff.u_sorted, unsorted_indices)
    ff.v = _take_along_axis2(ff.v_sorted, unsorted_indices)
    ff.w = _take_along_axis2(ff.w_sorted, unsorted_indices)

    ti_uns = _take_along_axis2(ff.turbulence_intensity_field_sorted, unsorted_indices)
    B, T, Ny, Nz = size(ti_uns)
    ti_avg = zeros(B, T, 1, 1)
    for b in 1:B, t in 1:T
        ti_avg[b, t, 1, 1] = mean(ti_uns[b, t, :, :])
    end
    ff.turbulence_intensity_field = ti_avg
    return ff
end
